File: BinarySearch/binary_search.cpp
#include <iostream>
#include <random>
#include <chrono>

const int SIZE = 100000;

int binary_search(int array[], int lower, int upper, int target)
{
    if(upper >= lower){
        int mid = (lower + upper) / 2;
        if(array[mid] == target){
            return mid;
        }else if(array[mid] < target){
            return binary_search(array, mid + 1, upper, target);
        }else{
            return binary_search(array, lower, mid - 1, target);
        }
    }
}

int main() {
    int array[SIZE];
    std::random_device rd;
    std::mt19937 mt(rd());
    std::uniform_real_distribution<double> dist(0, SIZE);
    for(int i = 0; i < SIZE; i++){
        array[i] = i;
    }
    int search = dist(mt);
    auto start = std::chrono::high_resolution_clock::now();
    auto finish = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(finish - start);
    int index = binary_search(array, 0, SIZE, search);
    std::cout << "Found target at index " << index << std::endl << "\tTook " << duration.count() << " ms";
    return 0;
}

File: BinarySearch/binary_search.rs
extern crate rand;

use std::time::{Instant};
use rand::Rng;

fn make_vec(size: u32) -> Vec<i32> {
    let mut array: Vec<i32> = Vec::new();
    for i in 0..size {
        array.push(i as i32);
    }
    return array
}

fn binary_search(array: Vec<i32>, lower: i32, upper: i32, target: i32) -> i32{
    if upper >= lower {
        let mid = (lower + upper) / 2;
        if array[mid as usize] == target { return mid }
        else if array[mid as usize] < target {
            return binary_search(array, mid + 1, upper, target);
        }
        else {
            return binary_search(array, lower, mid - 1, target);
        }
    }
    return -1
}

fn main() {
    let size: i32 = 100000000;
    let array = make_vec((size) as u32);
    let search = rand::thread_rng().gen_range(0, size - 1) ;
    let now = Instant::now();
    let index = binary_search(array, 0, size, search);
    let elapsed = now.elapsed().as_millis();
    println!("Found target at index {}", index);
    println!("\tTook {} ms", elapsed);
}
